aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+    /// Failed to create a kernel symbol resolver
+    #[error("Error creating KernelSymbolResolver: {error}")]
+    KernelSymbolResolverBuildError {
+        #[source]
+        /// The original io::Error
+        error: io::Error,
+    },
+}
+
+/// Stores all information that was resolved for a specific symbol
+#[derive(Debug, Clone)]
+pub struct SymbolInfo {
+    /// Virtual memory address of a certain function
+    virtual_address: u64,
+    /// Address of a function inside an object. `None` if it was not resolved
+    object_address: Option<u64>,
+    /// The PID of the process in user space
+    /// `None` is the symbol if the address is a kernel function
+    process_id: Option<u32>,
+    /// The resolved function name
+    function_name: Option<String>,
+    /// Path to the object which defines the symbol
+    object_path: Option<PathBuf>,
+}
+
+impl SymbolInfo {
+    /// Creates a `SymbolInfo` instance for a kernel-space function which hasn't been resolved yet
+    pub fn unresolved_kernel(address: u64) -> Self {
+        Self {
+            virtual_address: address,
+            object_address: Some(address),
+            function_name: None,
+            object_path: None,
+            process_id: None,
+        }
+    }
+
+    /// Creates a `SymbolInfo` instance for a user-space function which hasn't been resolved yet
+    pub fn unresolved_user(process_id: u32, address: u64) -> Self {
+        Self {
+            virtual_address: address,
+            object_address: None,
+            function_name: None,
+            object_path: None,
+            process_id: Some(process_id),
+        }
+    }
+
+    /// Whether the symbol refers to a kernel-space function
+    pub fn is_kernel(&self) -> bool {
+        self.process_id.is_none()
+    }
+
+    /// Whether the symbol refers to a user-space function
+    pub fn is_user(&self) -> bool {
+        self.process_id.is_some()
+    }
+
+    /// Returns the ID of the process for which the symbol was recorded
+    pub fn process_id(&self) -> Option<u32> {
+        self.process_id
+    }
+
+    /// Virtual memory address of a certain function
+    pub fn virtual_address(&self) -> u64 {
+        self.virtual_address
+    }
+
+    /// Address of a function inside an object. `None` if it was not resolved
+    pub fn object_address(&self) -> Option<u64> {
+        self.object_address
+    }
+
+    /// Path to the object which defines the symbol
+    pub fn object_path(&self) -> Option<&Path> {
+        self.object_path.as_deref()
+    }
+
+    /// The resolved function name
+    pub fn function_name(&self) -> Option<&str> {
+        self.function_name.as_deref()
+    }
+}
+
+/// Resolves a symbol based on it's address.
+///
+/// See [`DefaultResolver`] for exemplaric usage.
+pub trait SymbolResolver {
+    /// Resolves a symbol based on it's address
+    fn resolve(&self, symbol: &mut SymbolInfo);
+}
+
+/// A resolver for kernel symbols
+pub struct KernelSymbolResolver {
```

Comment:
Once `StackTrace::resolve` is removed, this resolver can also avoid `kernel_symbols()`  being a public function, since the resolver should be able to fulfill the whole job.